Shuttle timers

graevy opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
- fixed an oversight with TextScreenVisuals that caused it to display e.g. 1m60s
- created the `ShuttleTimer` prototype, which should automatically sync with parent shuttles when placed on them
- created `RemoteShuttleTimer` prototypes `EvacShuttleTimer` and `ArrivalsShuttleTimer` which should automatically sync with their respective shuttles, as well as track per-map ETA, when placed anywhere outside of them
- to avoid strict inheritance, duplicated much of the `SignalTimer` code into `Content.Server/TextScreen`. with a maintainer's blessing I would like to deduplicate signaltimer screens into textscreen composites

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
Implements [this feature](https://github.com/space-wizards/space-station-14/issues/18249) requested by @Emisse 

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Shuttles on creation now `EnsureComp<ShuttleTimerComponent>`, which contains hashset<entityuid>s `LocalScreens` and `RemoteScreens`. When a shuttle is created, any shuttletimers on it are added to `LocalScreens`, and `AllEntityQuery<RemoteShuttleTimerComponent>` filters from every placed remote timer (sorry; I couldn't come up with a better way to avoid local variable capture) using a bitmask to populate `RemoteScreens`. `ShuttleTimerComponent` is on both the timers and the shuttles; `RemoteShuttleTimerComponent` is only on the timers at the moment. Whenever a remote is created, it queries all shuttles to be added to their remotescreen hashset via the same bitmask.

There are some extra `ShuttleTimerEvent`s in the `ArrivalsSystem`, `RoundEndSystem`, `EmergencyShuttleSystem`, and `EmergencyShuttleSystem.Console` to handle timers at the start and end of each round, as well as useful timings specific to those systems (e.g. the emergency shuttle timers display the time until server restart after the round ends)

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

[I screencapped the timers I placed on maps here](https://mega.nz/file/KTAA1S7a#NGri9Jt9azpC3uFjYDp3-OH2C6QV46M_fGRlMwIiKfI), but removed them from the mapfiles after for mappers to place.

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

- add: Added shuttle timers and associated logic.

JoeHammad1844 commented:
video wont play for me

graevy commented:
> video wont play for me

can you download it?

JoeHammad1844 commented:
> can you download it?

didn't work

graevy commented:
> > can you download it?
> 
> didn't work

okay i just transcoded the shit out of it. it was x265

https://github.com/space-wizards/space-station-14/assets/51971268/2dec263c-58ea-4d43-b530-028082481957

graevy commented:
> > can you download it?
> 
> didn't work

https://github.com/space-wizards/space-station-14/assets/51971268/686422d3-e50b-426e-822a-36a7cdfa076f

compressed the shit out of it

JoeHammad1844 commented:
nice

metalgearsloth commented on Content.Server/Shuttles/Systems/ArrivalsSystem.cs:
```diff
@@ -47,6 +48,11 @@ public sealed class ArrivalsSystem : EntitySystem
     /// </summary>
     public bool Enabled { get; private set; }
 
+    public float Cooldown;
+    public float Startup = ShuttleSystem.DefaultStartupTime;
+    public float Travel = ShuttleSystem.DefaultTravelTime;
+    public float Arrival = ShuttleSystem.DefaultArrivalTime;
+    public float Docked;
```
Remove, data needs to be on components.

metalgearsloth commented on Content.Server/TextScreen/Systems/TextScreenSystem.cs:
```diff
@@ -0,0 +1,87 @@
+        SubscribeLocalEvent<TextScreenComponent, TextScreenTextEvent>(OnText);
+    }
+
+    private void OnInit(EntityUid uid, TextScreenComponent component, ComponentInit args)
+    {
+        _appearanceSystem.SetData(uid, TextScreenVisuals.ScreenText, component.Label);
+    }
```
Is this supposed to be MapInit?

metalgearsloth commented on Content.Server/TextScreen/Systems/TextScreenSystem.cs:
```diff
@@ -0,0 +1,87 @@
+        _appearanceSystem.SetData(uid, TextScreenVisuals.ScreenText, component.Label);
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+        UpdateTimer();
+    }
+
+    private void UpdateTimer()
+    {
+        var query = EntityQueryEnumerator<ActiveTextScreenTimerComponent, TextScreenComponent>();
+        while (query.MoveNext(out var uid, out var active, out var timer))
+        {
+            if (active.Remaining > _gameTiming.CurTime)
+                continue;
+
+            Finish(uid);
+
+            if (timer.DoneSound == null)
+                continue;
+            _audio.PlayPvs(timer.DoneSound, uid);
+        }
+    }
```
Just put the code in the update method.

metalgearsloth commented on Content.Server/TextScreen/Systems/TextScreenSystem.cs:
```diff
@@ -0,0 +1,87 @@
+
+using Robust.Shared.Timing;
+
+
+namespace Content.Server.TextScreen;
+
+public sealed class TextScreenSystem : EntitySystem
```
Systems need to be documented as this isn't intuitive.

metalgearsloth commented on Content.Server/TextScreen/Components/ActiveTextScreenTimerComponent.cs:
```diff
@@ -0,0 +1,10 @@
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom;
+
+namespace Content.Server.TextScreen.Components;
+
+[RegisterComponent]
+public sealed partial class ActiveTextScreenTimerComponent : Component
+{
+    [DataField("remaining", customTypeSerializer: typeof(TimeOffsetSerializer))]
+    public TimeSpan Remaining;
+}
```
Remove

metalgearsloth commented on Content.Server/Shuttles/Systems/ShuttleTimerSystem.cs:
```diff
@@ -0,0 +1,196 @@
+            }
+        }
+
+        // never got an answer figuring out how to do this. there's definitely a better pattern
+        /// <summary>
+        /// Raises a TEvent on each entity containing TComp.
+        /// </summary>
+        public void RaiseEventOnShuttles<TComp, TEvent>(ref TEvent ev)
+            where TComp : Component
+            where TEvent : struct
+        {
+            var query = AllEntityQuery<TComp>();
+            while (query.MoveNext(out var uid, out _))
+            {
+                RaiseLocalEvent(uid, ref ev);
+            }
+        }
```
Just add it to engine, also I don't think the event should have a struct constraint.

metalgearsloth commented on Content.Server/Shuttles/Systems/ShuttleSystem.cs:
```diff
@@ -106,6 +106,7 @@ private void OnGridInit(GridInitializeEvent ev)
             return;
 
         EntityManager.EnsureComponent<ShuttleComponent>(ev.EntityUid);
+        EntityManager.EnsureComponent<ShuttleTimerComponent>(ev.EntityUid);
```
Asteroids don't need shuttle timers.

metalgearsloth commented on Content.Server/Shuttles/Systems/ShuttleTimerSystem.cs:
```diff
@@ -0,0 +1,196 @@
+//      could directly access traitdicts at runtime?
+
+namespace Content.Server.Shuttles.Systems
+{
+    public struct MapFilter
+    {
+        public Func<EntityUid, EntityUid, bool> Comparator;
+        /// <summary>the MapUid that each timer's map is filtered against</summary>
+        public EntityUid MapUid;
+    }
+
```
This isn't documented and I'm kinda suss this needs to exist.

metalgearsloth commented on Content.Server/Shuttles/Components/ShuttleTimerComponent.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Content.Server.Shuttles.Components
+{
+    [RegisterComponent]
+    public sealed partial class ShuttleTimerComponent : Component
+    {
+        // these are hashsets because there's a lot of redundant pairing rn
+        [ViewVariables]
+        public HashSet<EntityUid> LocalScreens = new();
+
+        [ViewVariables]
+        public HashSet<EntityUid> RemoteScreens = new();
+
+        [ViewVariables]
+        public TimeSpan? Duration;
+
```
These need to be datafields (though really it should be using device networks).

metalgearsloth commented on Content.Server/Shuttles/Components/RemoteShuttleTimerComponent.cs:
```diff
@@ -0,0 +1,26 @@
+        [DataField("pairWith"), ViewVariables]
+        public RemoteShuttleTimerMask PairWith;
+    }
+
+    [Flags]
+    public enum RemoteShuttleTimerMask : byte
+    {
+        None = 0,
+        Emergency = 1 << 0,
+        Arrivals = 1 << 1,
+        Cargo = 1 << 2,
+        Salvage = 1 << 3,
+    }
```
Remove

metalgearsloth commented on Content.Server/Shuttles/Components/EmergencyShuttleComponent.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Content.Server.Shuttles.Components;
+
+[RegisterComponent]
+public sealed partial class EmergencyShuttleComponent : Component
+{
```
Document

metalgearsloth reviewed (changes requested):
Most of your systems and events need documenting as it's not clear when for example "TextScreenTextEvent" is raised.

Also this should be using device networks rather than its own separate pairing system.

graevy commented:
hi @metalgearsloth, i took a break and came back and addressed your feedback. it is much cleaner. i submitted [this pr](https://github.com/space-wizards/space-station-14/pull/21327) to help achieve a subnet pattern [here](https://github.com/space-wizards/space-station-14/blob/89a49440831aab3c50e72a9fb7dd7148362811f1/Content.Server/Shuttles/Systems/ShuttleTimerSystem.cs#L32)

- all pairing is done with device networks
- subnets for each timer proto (e.g. arrivals on terminal vs on station vs on shuttle) are still achieved by comparing `MapUid`s, but this time only in the network event handler
- `ShuttleTimer` and `TextScreenTimer` components/systems/events are now the only additions (outside of the protos)
- i added the timers to the shuttles/terminal/devmap in spots i thought worked. i wasn't sure if that was a good idea

i'm gonna go to sleep and figure out this awful commit history tomorrow.

UbaserB commented:
yeah its a bit weird it uses the brig timer sprite rn tbh

graevy commented:
alright, i'll grab another texture and graft the textscreensystem onto it as best i can

graevy commented:
alright @BolloTea @UbaserB how do we feel about the sprites now:

![dotnet_oS6RuPBu1D](https://github.com/space-wizards/space-station-14/assets/51971268/252f0adb-23c7-4035-b517-6f0b45fc1d49)

UbaserB commented:
![Uploading IMG_0959.png…]()

BolloTea commented:
Looks much better.  But, can I ask you one small selfish request?  Wouldn't it be too difficult to make a small abbreviated inscription on the screen itself (time until arrival and a separate time before departure of the shuttle)?

graevy commented:
@BolloTea "ETA"/"ETD" in a row above the timer?

graevy commented:
completely refactored the client `TextScreenSystem` to:
- split the text/timer state behavior between `TextScreenVisualsComponent` and `TextScreenTimerComponent`
- use a `string?[]` instead of a `string` for text, to accommodate multiple text rows
- center rendered text per-row

did a minor `signaltimersystem` refactor to accommodate this

removed the server textscreensystem and events; using the appearance system instead

generalized `shuttletimers` into `screens` so they can extend to the hallway screens from ss13

added the screens to the arrivals/emergency shuttles, and the terminal/centcomm/dev maps

renamed `textscreen` (the brig timer sprites) to `signalscreen` to make the distinction clearer

documented all the new methods and added an overview to the refactored textscreensystem

metalgearsloth commented on Content.Server/Shuttles/Systems/ShuttleTimerSystem.cs:
```diff
@@ -0,0 +1,98 @@
+using Robust.Shared.Timing;
+
+
+namespace Content.Server.Shuttles.Systems
+{
```
file-scoped namespaces please.

metalgearsloth commented on Content.Server/Shuttles/Systems/ShuttleTimerSystem.cs:
```diff
@@ -0,0 +1,98 @@
+            _appearanceSystem.SetData(uid, TextScreenVisuals.ScreenText, text);
+        }
+
+        /// <summary>
+        /// Helper method for <see cref="RoundEndSystem"/> and <see cref="EmergencyShuttleSystem"/> to get the MapUids of
+        /// the `eshuttle`, `centcomm`, and the best-guess `station` (using <see cref="StationSystem.GetLargestGrid(StationDataComponent)"/>)
+        /// </summary>
+        /// <returns>{"eshuttle":uid?, "centcomm":uid?, "station":uid?}</returns>
+        public Dictionary<string, EntityUid?> GetEscapeMaps()
+        {
+            AllEntityQuery<StationEmergencyShuttleComponent>().MoveNext(out var station, out var eshuttleComp);
+            var targetGrid = _stationSystem.GetLargestGrid(Comp<StationDataComponent>(station));
+            var stationMap = targetGrid == null ? null : Transform(targetGrid.Value).MapUid;
+
+            var eshuttleMap = eshuttleComp?.EmergencyShuttle;
+
+            AllEntityQuery<StationCentcommComponent>().MoveNext(out var centcomm);
+            EntityUid? centcommMap = centcomm == null ? null : _mapManager.GetMapEntityId(centcomm.MapId);
+
+            return new Dictionary<string, EntityUid?>
+            {
+                {"eshuttle", eshuttleMap}, {"centcomm", centcommMap}, {"station", stationMap}
+            };
+        }
```
This method feels weird but I can't articular how to make it better so it's fine as is.

metalgearsloth commented on Content.Server/Shuttles/Systems/ShuttleTimerSystem.cs:
```diff
@@ -0,0 +1,98 @@
+        /// <returns>{"eshuttle":uid?, "centcomm":uid?, "station":uid?}</returns>
+        public Dictionary<string, EntityUid?> GetEscapeMaps()
+        {
+            AllEntityQuery<StationEmergencyShuttleComponent>().MoveNext(out var station, out var eshuttleComp);
+            var targetGrid = _stationSystem.GetLargestGrid(Comp<StationDataComponent>(station));
```
You already have the query here just add StationDataComponent as another generic.

metalgearsloth commented on Content.Server/Shuttles/Systems/EmergencyShuttleSystem.cs:
```diff
@@ -206,6 +268,24 @@ public void CallEmergencyShuttle(EntityUid stationUid, StationEmergencyShuttleCo
+            // shuttle timers
+            var time = TimeSpan.FromSeconds(_consoleAccumulator);
+            if (TryComp<DeviceNetworkComponent>(stationShuttle.EmergencyShuttle.Value, out var netComp))
+            {
+                AllEntityQuery<StationCentcommComponent>().MoveNext(out var centcomm);
```
Just dump this on a method instead IMO, like "GetCentcomm" or something so at least we only have 1 spot to update if we have to change it.
